compiler/src/dotty/tools/MainGenericCompiler.scala

```diff
@@ -0,0 +1,195 @@
+
+  def withColors: CompileSettings =
+    this.copy(colors = true)
+
+  def withNoColors: CompileSettings =
+    this.copy(colors = false)
+}
+
+object MainGenericCompiler {
+
+  val classpathSeparator = File.pathSeparator
+
+  @sharable val javaOption = raw"""-J(.*)""".r
+  @sharable val javaPropOption = raw"""-D(.+?)=(.?)""".r
+  @tailrec
+  def process(args: List[String], settings: CompileSettings): CompileSettings = args match
+    case Nil =>
+      settings
+    case "--" :: tail =>
+      process(Nil, settings.withResidualArgs(tail.toList*))
+    case ("-v" | "-verbose" | "--verbose") :: tail =>
+      process(tail, settings.withScalaArgs("-verbose"))
+    case ("-q" | "-quiet") :: tail =>
+      process(tail, settings.withQuiet)
+    case "-Oshort" :: tail =>
+      process(tail, settings.withJavaArgs("-XX:+TieredCompilation", "-XX:TieredStopAtLevel=1"))
+    case "-repl" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Repl))
+    case "-script" :: targetScript :: tail =>
+      process(Nil, settings
+        .withCompileMode(CompileMode.Script)
+        .withJavaProps("script.path" -> targetScript)
+        .withTargetScript(targetScript)
+        .withScriptArgs(tail.toList*))
+    case "-compile" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Compile))
+    case "-decompile" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Decompile))
+    case "-print-tasty" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.PrintTasty))
+    case "-run" :: tail =>
+      process(tail, settings.withCompileMode(CompileMode.Run))
+    case "-colors" :: tail =>
+      process(tail, settings.withColors)
+    case "-no-colors" :: tail =>
+      process(tail, settings.withNoColors)
+    case "-with-compiler" :: tail =>
+      process(tail, settings.withCompiler)
+    case ("-cp" | "-classpath" | "--class-path") :: cp :: tail =>
+      val cpEntries = cp.split(classpathSeparator).toList
+      val singleEntryClasspath: Boolean = cpEntries.sizeIs == 1
+      val globdir: String = if singleEntryClasspath then cp.replaceAll("[\\\\/][^\\\\/]*$", "") else "" // slash/backslash agnostic
+      def validGlobbedJar(s: String): Boolean = s.startsWith(globdir) && ((s.toLowerCase.endsWith(".jar") || s.toLowerCase.endsWith(".zip")))
+      val (tailargs, newEntries) = if singleEntryClasspath && validGlobbedJar(cpEntries.head) then
+        // reassemble globbed wildcard classpath
+        // globdir is wildcard directory for globbed jar files, reconstruct the intended classpath
+        val cpJars = tail.takeWhile( f => validGlobbedJar(f) )
+        val remainingArgs = tail.drop(cpJars.size)
+        (remainingArgs, cpEntries ++ cpJars)
+      else
+        (tail, cpEntries)
+
+      process(tailargs, settings.copy(classPath = settings.classPath ++ newEntries.filter(_.nonEmpty)))
+    case (o @ javaOption(stripped)) :: tail =>
+      process(tail, settings.withJavaArgs(stripped))
```

Comment:
```suggestion
    case javaOption(stripped) :: tail =>
      process(tail, settings.withJavaArgs(stripped))
```

Reply:
This was also copied from MainGenericRunner